src/protocolhandler.ts

```diff
@@ -192,10 +196,48 @@ export class ProtocolHandler extends EventEmitter {
         this.outbound.reset();
     };
 
+    drain(): Promise<any> {
+        if (this.closed) {
+            return Promise.reject(NatsError.errorForCode(ErrorCode.CONN_CLOSED));
+        }
+        if (this.draining) {
+            return Promise.reject(NatsError.errorForCode(ErrorCode.CONN_DRAINING));
+        }
+        this.draining = true;
+
+        let subs = this.subscriptions.all();
+        let promises: Promise<SubEvent>[] = [];
+        let mux = this.subscriptions.mux ? this.subscriptions.mux.sid : 0;
```

Comment:
Not sure you should treat this one differently. Need to be discussed.

Reply:
several - if the client needs to do reqrep as part of processing during drain, the mux is required.

I think the main issue is that a client that is doing req/rep as part of processing subscription events likely needs to do multiple connections. That would solve the issue. The other alternative which I think it is the correct way, is for the client to drain specific subscriptions. This enables the client to do whatever it needs to do while insuring that the subscription messages are processed.